### PR TITLE
Feature/108 get generic events

### PR DESCRIPTION
**Pull request by b-brodie-digdir**
Added new endpoint for retrieving cloud events from the events database (Postgresql).

## Description
Migrated existing functionality from App Events and refactored to support generic cloud events.

## Related Issue(s)
- #108 

## Verification
- [x] **Your** code builds clean without any errors or warnings
- [ ] Manual testing done (required)
- [x] Relevant automated test added (if you find this hard, leave it and we'll help out)
- [x] All tests run green

## Documentation
- [ ] User documentation is updated - see `feature/events` branch in the `altinn-studio-docs` repo.

### CODE REVIEW

File: src/Events/Controllers/EventsController.cs [hunk 1]
```diff
@@ -63,6 +77,97 @@ public async Task<ActionResult<string>> Post([FromBody] CloudEvent cloudEvent)
...
+        /// A list of the sources to include</param>
+        /// <param name="type" example="[&quot;app.instance.created&quot;, &quot;app.instance.process.completed&quot;]">
+        /// A list of the event types to include</param>
+        /// <param name="subject">Optional filter by subject. Only exact matches will be returned.</param>
+        /// <param name="size">The maximum number of events to include in the response.</param>
+        [HttpGet()]
+        [Consumes("application/json")]
+        [ProducesResponseType(StatusCodes.Status200OK)]
+        [ProducesResponseType(StatusCodes.Status400BadRequest)]
+        [Authorize(Policy = AuthorizationConstants.SCOPE_EVENTS_SUBSCRIBE)]
```
Comment (acn-sbuad) [inline src/Events/Controllers/EventsController.cs:95]: both publish and subscribe should give access to this point I believe.

File: .github/workflows/use-case-ATX.yaml [hunk 2]
```diff
@@ -20,7 +20,7 @@ jobs:
     - name: Run events use case tests
       run:  |
               cd test/k6
-              docker-compose run k6 run /src/tests/events.js -e env=${{ vars.ENV }} -e tokenGeneratorUserName=${{ secrets.TOKENGENERATOR_USERNAME }} -e tokenGeneratorUserPwd=${{ secrets.TOKENGENERATOR_USERPASSWORD }}
+              docker-compose run k6 run /src/tests/events/post.js -e env=${{ vars.ENV }} -e tokenGeneratorUserName=${{ secrets.TOKENGENERATOR_USERNAME }} -e tokenGeneratorUserPwd=${{ secrets.TOKENGENERATOR_USERPASSWORD }}
```
Comment (acn-sbuad) [inline .github/workflows/use-case-ATX.yaml:23]: use-case tt and prod also refere to this file

File: src/Events/Repository/ICloudEventRepository.cs [hunk 3]
```diff
@@ -28,5 +28,10 @@ public interface ICloudEventRepository
         /// Calls a function to retrieve app cloud events based on query params
         /// </summary>
         Task<List<CloudEvent>> GetAppEvents(string after, DateTime? from, DateTime? to, string subject, List<string> source, List<string> type, int size);
+
+        /// <summary>
+        /// Calls a function to retrieve app cloud events based on query params
```
Comment (acn-sbuad) [inline src/Events/Repository/ICloudEventRepository.cs:33]:
```suggestion
        /// Calls a function to retrieve cloud events based on query params
```

File: test/Altinn.Platform.Events.Tests/TestingServices/EventsServiceTest.cs [hunk 4]
```diff
@@ -249,6 +249,141 @@ public async Task GetAppEvents_NoSubjectInfoAvailable_RegisterLookupIsNotCalled(
             registerMock.Verify(r => r.PartyLookup(It.Is<string>(s => string.IsNullOrEmpty(s)), It.Is<string>(s => string.IsNullOrEmpty(s))), Times.Never);
         }
 
+        #region Generic events
+
+        /// <summary>
+        /// Scenario:
+        ///   Get instances based on from and party Id
```
Comment (acn-sbuad) [inline test/Altinn.Platform.Events.Tests/TestingServices/EventsServiceTest.cs:256]:
```suggestion
        ///   Get instances based on after and party Id
```

File: src/Events/Controllers/EventsController.cs [hunk 5]
```diff
@@ -63,6 +77,97 @@ public async Task<ActionResult<string>> Post([FromBody] CloudEvent cloudEvent)
...
+        [Produces("application/cloudevents+json")]
+        public async Task<ActionResult<List<CloudEvent>>> Get(
+            [FromQuery] string after,
+            [FromQuery] List<string> source,
+            [FromQuery] List<string> type,
+            [FromHeader] string subject,
+            [FromQuery] int size = 50)
+        {
+            (bool isValid, string errorMessage) = ValidateQueryParams(after, size);
```
Comment (SandGrainOne) [inline src/Events/Controllers/EventsController.cs:104]: How do we validate that a source value has been provided?

File: test/k6/src/tests/events/get.js [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+  response = eventsApi.getCloudEvents("0", 
+    data.cloudEvent.source,
+    data.cloudEvent.type,
+    data.cloudEvent.subject,
+    10,
+    data.token
+  );
+
+  success = check(response, {
```
Comment (acn-sbuad) [inline test/k6/src/tests/events/get.js:54]: should we check that the list has content as well?

File: test/k6/src/tests/events/get.js [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+  if (data.runFullTestSet) {
+    TC01_GetAllEvents(data);
+    
+  } else {
+    // Limited test set for use case tests
+    TC01_GetAllEvents(data);
+  }
+}
+
+function removePropFromCloudEvent(cloudEvent, propertyname) {
```
Comment (acn-sbuad) [inline test/k6/src/tests/events/get.js:75]: don't think you need this function for the `get` tests

File: src/Events/Controllers/EventsController.cs [hunk 8]
```diff
@@ -20,16 +28,23 @@ namespace Altinn.Platform.Events.Controllers
...
+        public EventsController(
+            IEventsService events,
+            ILogger<EventsController> logger,
+            IOptions<GeneralSettings> settings)
         {
-            _events = events;
+            _eventsService = events;
+            _logger = logger;
             _settings = settings.Value;
+            _eventsBaseUri = $"https://platform.{settings.Value.Hostname}";
```
Comment (acn-sbuad) [inline src/Events/Controllers/EventsController.cs:47]: this is now set both for app and eventscontroller. How about generating a seperate config value for this baseUri ?

File: src/Events/Controllers/EventsController.cs (hunk 5 above)
Comment (b-brodie-digdir) [reply src/Events/Controllers/EventsController.cs]: Validation of source added in a more recent commit.

File: test/k6/src/tests/events/get.js (hunk 6 above)
Comment (b-brodie-digdir) [reply test/k6/src/tests/events/get.js:55]: Good idea

File: test/k6/src/tests/events/get.js (hunk 7 above)
Comment (b-brodie-digdir) [reply test/k6/src/tests/events/get.js]: Cleaned-up

File: src/Events/Controllers/EventsController.cs (hunk 8 above)
Comment (b-brodie-digdir) [reply src/Events/Controllers/EventsController.cs]: I pulled out the whole uri and then we don't need the `Hostname` setting. Should set this with a known good value for our environments.

File: test/k6/src/tests/events/get.js [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+  var data = {
+    runFullTestSet: runFullTestSet,
+    token: token,
+    cloudEvent: cloudEvent,
+  };
+
+  return data;
+}
+
+// 01 - POST valid cloud event with all parameters
```
Comment (SandGrainOne) [inline test/k6/src/tests/events/get.js:40]: Should have been a GET?

File: test/Altinn.Platform.Events.Tests/appsettings.json [hunk 10]
```diff
@@ -3,7 +3,7 @@
     "EnableDBConnection": "false"
   },
   "GeneralSettings": {
-    "Hostname": "localhost:5080",
+    "BaseUri": "https://platform.localhost:5080",
```
Comment (SandGrainOne) [inline test/Altinn.Platform.Events.Tests/appsettings.json:6]:
```suggestion
    "BaseUri": "http://localhost:5080",
```

File: test/Altinn.Platform.Events.Tests/TestingServices/EventsServiceTest.cs [hunk 11]
```diff
@@ -249,6 +249,141 @@ public async Task GetAppEvents_NoSubjectInfoAvailable_RegisterLookupIsNotCalled(
...
+            // Arrange
+            Mock<IRegisterService> registerMock = new();
+
+            EventsService eventsService = GetEventsService(registerMock: registerMock);
+
+            // Act
+            List<CloudEvent> actual = await eventsService.GetEvents("1", new List<string>() { "https://ttd.apps.at22.altinn.cloud/ttd/app-test/" }, new List<string>() { }, null);
+
+            // Assert
+            registerMock.Verify(r => r.PartyLookup(It.Is<string>(s => string.IsNullOrEmpty(s)), It.Is<string>(s => string.IsNullOrEmpty(s))), Times.Never);
```
Comment (SandGrainOne) [inline test/Altinn.Platform.Events.Tests/TestingServices/EventsServiceTest.cs:382]:
```suggestion
            registerMock.Verify(r => r.PartyLookup(It.IsAny<string>(), It.IsAny<string>()), Times.Never);
```

File: test/Altinn.Platform.Events.Tests/TestingControllers/EventsControllerTests.cs [hunk 12]
```diff
@@ -123,6 +128,212 @@ public async Task Post_ExternalEventsDisabled_NotFoundResponse()
...
+                HttpRequestMessage httpRequestMessage = new HttpRequestMessage(HttpMethod.Get, requestUri);
+
+                // Act
+                HttpResponseMessage response = await client.SendAsync(httpRequestMessage);
+                string responseString = await response.Content.ReadAsStringAsync();
+                List<CloudEvent> actual = JsonSerializer.Deserialize<List<CloudEvent>>(responseString);
+
+                // Assert
+                Assert.Equal(HttpStatusCode.OK, response.StatusCode);
+                Assert.Equal(expectedCount, actual.Count);
```
Comment (SandGrainOne) [inline test/Altinn.Platform.Events.Tests/TestingControllers/EventsControllerTests.cs:213]: Just a heads up. The Count is indirectly produced by a mock. Not by the code under test.

File: src/Events/appsettings.json [hunk 13]
```diff
@@ -3,12 +3,12 @@
...
     "WorkspacePath": "Migration",
     "AdminConnectionString": "Host=localhost;Port=5432;Username=platform_events_admin;Password={0};Database=eventsdb",
-    "ConnectionString": "Host=localhost;Port=5432;Username=platform_events;Password={0};Database=eventsdb",
+    "ConnectionString": "Host=localhost;Port=5432;Username=platform_events;Password={0};Database=eventsdb;Include Error Detail=True",
     "EventsDbAdminPwd": "Password",
     "EventsDbPwd": "Password"
   },
   "GeneralSettings": {
-    "Hostname": "localhost:5080",
+    "BaseUri": "https://platform.localhost:5080",
```
Comment (SandGrainOne) [inline src/Events/appsettings.json:11]:
```suggestion
    "BaseUri": "http://localhost:5080",
```
I didn't test this, but I think this is closer to the correct uri.

**Review by SandGrainOne — APPROVED** [review]
Commented on a few things, but this looks okay to me.

File: test/k6/src/tests/events/get.js (hunk 9 above)
Comment (b-brodie-digdir) [reply test/k6/src/tests/events/get.js:40]: Good catch

File: src/Events/Services/Interfaces/IEventsService.cs [hunk 14]
```diff
@@ -38,5 +38,10 @@ public interface IEventsService
         /// Gets list of cloud events based on query params
         /// </summary>
         Task<List<CloudEvent>> GetAppEvents(string after, DateTime? from, DateTime? to, int partyId, List<string> source, List<string> type, string unit, string person, int size = 50);
+
+        /// <summary>
+        /// Gets list of cloud events based on query params
+        /// </summary>
+        Task<List<CloudEvent>> GetEvents(string after, List<string> source, List<string> type, string subject, int size = 50);
```
[the remark this answers is not in the document]
Comment (b-brodie-digdir) [reply src/Events/Services/Interfaces/IEventsService.cs:45]: It does in fact, and generally more readable than deeper in the tree. 
It is not advised to mix default values on interfaces and concrete implementations though, so I need to consider this more before introducing a default here.

File: test/Altinn.Platform.Events.Tests/TestingControllers/EventsControllerTests.cs (hunk 12 above)
Comment (b-brodie-digdir) [reply test/Altinn.Platform.Events.Tests/TestingControllers/EventsControllerTests.cs:213]: Good point, no sense in cluttering our assert list.

File: src/Events/appsettings.json (hunk 13 above)
Comment (b-brodie-digdir) [reply src/Events/appsettings.json:11]: I was considering doing that change, so now I have to :)

File: test/Altinn.Platform.Events.Tests/TestingServices/EventsServiceTest.cs (hunk 11 above)
Comment (b-brodie-digdir) [reply test/Altinn.Platform.Events.Tests/TestingServices/EventsServiceTest.cs:382]: Very hard.

File: test/Altinn.Platform.Events.Tests/appsettings.json (hunk 10 above)
Comment (b-brodie-digdir) [reply test/Altinn.Platform.Events.Tests/appsettings.json:6]: Fixed.

File: src/Events/Services/Interfaces/IEventsService.cs (hunk 14 above)
[the remark this answers is not in the document]
Comment (b-brodie-digdir) [reply src/Events/Services/Interfaces/IEventsService.cs]: It does in fact, and generally more readable than deeper in the tree.
It is not advised to mix default values on interfaces and concrete implementations though, so I need to consider this more before introducing a default here.
We have the same default value specified in three locations, I think perhaps we should pick one or two:
- controller parameter
- service interface
- service implementation
- postgres function

File: test/Altinn.Platform.Events.Tests/appsettings.json (hunk 10 above)
Comment (b-brodie-digdir) [reply test/Altinn.Platform.Events.Tests/appsettings.json]: I tried replying to these comments using the new (built-in?) support in VS Code. Didn't seem to work quite right. 

I've updated all the tests to use this Uri

File: test/k6/src/tests/events/get.js (hunk 9 above)
Comment (b-brodie-digdir) [reply test/k6/src/tests/events/get.js]: fixed.

File: test/Altinn.Platform.Events.Tests/TestingServices/EventsServiceTest.cs (hunk 11 above)
Comment (b-brodie-digdir) [reply test/Altinn.Platform.Events.Tests/TestingServices/EventsServiceTest.cs]: Added.

File: .github/workflows/use-case-ATX.yaml (hunk 2 above)
Comment (acn-sbuad) [inline .github/workflows/use-case-ATX.yaml:23]: Get tests should also be included in the use-case run I guess

File: test/Altinn.Platform.Events.Tests/TestingControllers/AppControllerTests.cs [hunk 15]
```diff
@@ -413,7 +413,7 @@ public async void GetForOrg_ValidRequest_ReturnsListOfEventsAndNextUrl()
             {
                 // Arrange
                 string requestUri = $"{BasePath}/app/ttd/endring-av-navn-v2?from=2020-01-01Z&party=1337";
-                string expectedNext = $"https://platform.localhost:5080/events/api/v1/app/ttd/endring-av-navn-v2?after=e31dbb11-2208-4dda-a549-92a0db8c8808&from=2020-01-01Z&party=1337";
+                string expectedNext = $"http://localhost:5080/events/api/v1/app/ttd/endring-av-navn-v2?after=e31dbb11-2208-4dda-a549-92a0db8c8808&from=2020-01-01Z&party=1337";
```
Comment (acn-sbuad) [inline test/Altinn.Platform.Events.Tests/TestingControllers/AppControllerTests.cs:416]: Not quite sure about this change here. Any particular reason you changed it? As requests are converted to the internal address using http when request hits the pod, we want to confirm that the scheme is still https then the next url is returned to the end user

**Review by acn-sbuad — CHANGES_REQUESTED** [review]
Some concerns with changes in the unit test w.r.t nextUrl

File: test/k6/src/tests/events/get.js [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+    data.cloudEvent.type,
+    data.cloudEvent.subject,
+    10,
+    data.token
+  );
+
+
+  success = check(response, {
+    "GET all cloud events: status is 200": (r) => r.status === 200,
+    "GET all cloud events: at least 1 cloud event returned": (r) =>
```
Comment (acn-sbuad) [inline test/k6/src/tests/events/get.js:57]: What about verifying the next url as well? At least the scheme returned. Ref the comment in the unit tests

File: test/Altinn.Platform.Events.Tests/TestingControllers/AppControllerTests.cs (hunk 15 above)
Comment (SandGrainOne) [reply test/Altinn.Platform.Events.Tests/TestingControllers/AppControllerTests.cs:416]: Not entirely correct. The code no longer set the scheme, but picks scheme from a config. If we want to test this specifically we could add gibberish to the config to confirm that the value is used.

Comment (acn-sbuad) [reply test/Altinn.Platform.Events.Tests/TestingControllers/AppControllerTests.cs:416]: Had a chat with @b-brodie, we need to update environment values in the helm chart to reflect this change before this is merged then

**Comment by b-brodie-digdir** [comment]
PR for config changes: https://dev.azure.com/brreg/altinn-studio-ops/_git/altinn-studio-ops/pullrequest/610

File: .github/workflows/use-case-ATX.yaml (hunk 2 above)
Comment (b-brodie-digdir) [reply .github/workflows/use-case-ATX.yaml:25]: Added.

File: test/Altinn.Platform.Events.Tests/TestingControllers/AppControllerTests.cs (hunk 15 above)
Comment (b-brodie-digdir) [reply test/Altinn.Platform.Events.Tests/TestingControllers/AppControllerTests.cs:416]: I've added values for the new `BaseUri` config, for all environments.

**Review by acn-sbuad — APPROVED** [review]

**Pull request merged**